History page

jaminko opened the pull request.

varajan committed:
Enable history page

Commit:
- Deleted unused method
- Add History Page
- Add HistoryLinkTest

Commit:
- Add Calculator Link Test
- Add Title Test
- Add Last Calculation History Test
- Add Total Number Of Clculations Test
- Add Clear ButtonTest

Commit:
- Code refactoring
- Add Column Signatures Test

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,123 @@
+        [Test]
+        public void TitleTest()
+        {
+            // Arrange
+            string historyPageTitle = "History";
+
+            // Assert
+            Assert.AreEqual(historyPageTitle, driver.Title, "Incorrect title");
```
It could be done all in one line:
```suggestion
            Assert.AreEqual("History", driver.Title, "Incorrect title");
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,123 @@
+            Assert.AreEqual(historyPageTitle, driver.Title, "Incorrect title");
+        }
+
+        [TestCase("177", "64", "25")]
+        public void LastCalculationHistoryTest(string depositAmount, string interestRate, string investmentTerm)
+        {
+            // Arrange
+            var historyPage = new HistoryPage(driver);
+            var calculatorPage = new CalculatorPage(driver);
```
You can move the variables to the test body, if there is only one test-case:
```suggestion
        [Test]
        public void LastCalculationHistoryTest()
        {
            // Arrange
            var depositAmount = "177";
            var interestRate = "64";
            var investmentTerm = "25";
            var historyPage = new HistoryPage(driver);
            var calculatorPage = new CalculatorPage(driver);
```

varajan commented on DepositeCalcTests/Pages/HistoryPage.cs:
```diff
@@ -0,0 +1,71 @@
+            return new CalculatorPage(driver);
+        }
+
+        public HistoryPage Clear()
+        {
+            ClearBtn.Click();
+            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(_ => ClearBtn.Enabled);
+            return new HistoryPage(driver);
+        }
```
This action doesn't return a new page, so it could be `void`. I prefer this solution.
```suggestion
        public void Clear()
        {
            ClearBtn.Click();
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(_ => ClearBtn.Enabled);
        }
```

But, if you want the method to return the current page, you can return `this`.
```suggestion
        public HistoryPage Clear()
        {
            ClearBtn.Click();
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(_ => ClearBtn.Enabled);
            return this;
        }
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,123 @@
+{
+    internal class HistoryPageTests
+    {
+        private IWebDriver driver;
+
+        [SetUp]
+        public void Setup()
+        {
+            ChromeOptions options = new ChromeOptions { AcceptInsecureCertificates = true };
+            driver = new ChromeDriver(options);
+            driver.Manage().Timeouts().PageLoad = TimeSpan.FromSeconds(30);
+            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(5);
+            driver.Url = "https://localhost:5001/History";
+        }
+
```
Each your test begins with `var historyPage = new HistoryPage(driver);`, so you can declare global variable `historyPage` in the class and move its initialization to the `Setup`.

```suggestion
        private IWebDriver driver;
        private HistoryPage historyPage;

        [SetUp]
        public void Setup()
        {
            ChromeOptions options = new ChromeOptions { AcceptInsecureCertificates = true };
            driver = new ChromeDriver(options);
            driver.Manage().Timeouts().PageLoad = TimeSpan.FromSeconds(30);
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(5);
            driver.Url = "https://localhost:5001/History";
            historyPage = new HistoryPage(driver);
        }

```

varajan commented on DepositeCalcTests/Pages/HistoryPage.cs:
```diff
@@ -0,0 +1,71 @@
+            return new HistoryPage(driver);
+        }
+
+        public string HistoryTableRow1Column1 => HistoryTableRow1Coloumn1Field.Text;
```
Good work, but it's better to use lists or arrays. E.g. to create such properties:
```
List<string> Headers => ...
List<string> LastRow => ...
List<List<string>> TableValues => ...
```

varajan committed:
#27 Fix page title

varajan committed:
#28 show 10 last history records

varajan committed:
#29 move column 'Interest' to the last position.

varajan committed:
#31 don't show % symbol in the table

Commit:
- Add BaseTest
- Add InitDriver method

Commit:
- Add WeitForReady method for History page
- Resolved conversations

Commit:
- Changed LastCalculationHistoryTest (using arrays)
- Changed ColumnSignaturesTest (using arrays)
- Code refactoring
- Resolved ALL conversations

varajan committed:
Try to hide debug messages

varajan committed:
Try to hide debug information messages

Commit:
- Add method for clearing mandatory text fields
- Add method for create ten random calculations (for History tests)
- Code refactoring

Commit:
Merge branch 'HistoryPage' of https://github.com/jaminko/web.test into HistoryPage

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,106 @@
+            driver.Url = "https://localhost:5001/Calculator";
+            calculatorPage.ClearingMandatoryTextFields();
+            calculatorPage.FinancialYear = "365";
+            calculatorPage.FillingMandatoryTextFields(depositAmount, interestRate, investmentTerm);
```
1. Let's create an Open() method and move `driver.Url = "https://localhost:5001/Calculator";` to it. So the test will look much better.
2. I think you don't need to 'clear fields' here.
```suggestion
            calculatorPage.Open();
            calculatorPage.FinancialYear = "365";
            calculatorPage.FillingMandatoryTextFields(depositAmount, interestRate, investmentTerm);
```

varajan replied:
Hard-code start date, and you can verify the whole table row.
```
    calculatorPage.StartDataYear = "2020";
    calculatorPage.StartDataMonth = "March";
    calculatorPage.StartDataDay = "1";
```

varajan commented on DepositeCalcTests/Tests/CalculatorPageTests.cs:
```diff
@@ -174,15 +145,22 @@ public void YearsFieldTests()
             Assert.IsTrue(settingsPage.IsOpened(), "Incorrect page");
         }
+
+        [Test]
+        public void HistoryLinkTest()
```
It is very good, that you update 'old' test classes with new tests.

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,106 @@
+                "365",
+                "231.17",
+                "108.17",
+            };
```
You can write lists in one line. In current case it should be better - it is a history row.
```suggestion
            var expectedLastCalculationValues = new[]  { "123.00", "100", "321", "365", "231.17", "108.17", };
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,106 @@
+                "To",
+                "Income",
+                "Interest",
+            };
```
```suggestion
            var expectedHeaders = new[] {  "Amount",  "%",  "Term", "Year",  "From", "To", "Income", "Interest",  };
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,106 @@
+            calculatorPage.ClearingMandatoryTextFields();
+            calculatorPage.FinancialYear = "365";
+            calculatorPage.FillingMandatoryTextFields(depositAmount, interestRate, investmentTerm);
+            calculatorPage.Calculate();
+            Thread.Sleep(1000); // doesn't work without this delay
+            calculatorPage.OpenHistory();
+            historyPage.WeitForReady();
```
It will work without sleep...
```suggestion
            calculatorPage.Calculate();
            calculatorPage.OpenHistory();
            historyPage.WeitForReady();
```

...but you should update the wait in `Calculate` method:
```
        public void Calculate()
        {
            CalculateBtn.Click();
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(_ => CalculateBtn.Enabled);
        }
```

varajan commented on DepositeCalcTests/Pages/CalculatorPage.cs:
```diff
@@ -112,5 +112,34 @@ public bool IsOpened()
+            return new HistoryPage(driver);
+        }
+
+        public void ClearingMandatoryTextFields()
```
I suggest to move these 'clears' to the 'FillingMandatoryTextFields' method:
```
        public void FillingMandatoryTextFields(string depositAmount, string interestRate, string investmentTerm)
        {
            DepositAmountFld.Clear();
            DepositAmountFld.SendKeys(depositAmount);
            RateOfInterestFld.Clear();
            RateOfInterestFld.SendKeys(interestRate);
            InvestmentTermFld.Clear();
            InvestmentTermFld.SendKeys(investmentTerm);
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(_ => CalculateBtn.GetAttribute("disable") != string.Empty);
        }
```

varajan commented on DepositeCalcTests/Pages/CalculatorPage.cs:
```diff
@@ -112,5 +112,34 @@ public bool IsOpened()
+            Random random = new Random();
+            for (int i = 0; i < 10; i++)
+            {
+                int depositAmount = random.Next(1, 10000);
+                int interestRate = random.Next(1, 100);
+                int investmentTerm = random.Next(1, 365);
+                FillingMandatoryTextFields(depositAmount.ToString(), interestRate.ToString(), investmentTerm.ToString());
```
Move the whole code to the test. And you can hard-code the expected table values.
```suggestion
                FillingMandatoryTextFields("10000", $"{(10+i)}", "365");
                // StartDateYear = "2020";
                // StartDateMonth = "July";
                // StartDateDay = $"{(1+i)}"
```

Commit:
- Code refactoring
- Resolved ALL conversations

Commit:
- Add HistoryTableTest
- Delete LastCalculationHistoryTest (dublicated)
- Delete CreateTenRandomCalculations method (implement some part of this in HistoryTableTest)
- Add waiter for currency field
- Code refactoring

Commit:
- Delete WeitCurrencyFieldReady method

Commit:
- Add waiter for ValidLoginTest

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,103 @@
+        public void CalculatorLinkTest()
+        {
+            // Act
+            var calculatorPage = historyPage.Calculator();
```
Let's find better name for `Calculator` method, e.g. OpenCaluculator, ReturnToCalculator

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            // Act
+            var unclearedHistoryPage = historyPage.WeitForReady();
+            historyPage.Clear();
```
You should create at least one record for this test.

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,103 @@
+        public void TotalNumberOfClculationsTest()
+        {
+            // Assert
+            Assert.AreEqual(10, historyPage.HistoryTableNumberOfRows, "The History table doesn't contain the last ten calculations");
```
This test is dependent on `HistoryTableTest` test. If I run it separately, it will fail. Actually, this check is covered by that test, so you may just delete it.

varajan commented on DepositeCalcTests/Tests/LoginPageTests.cs:
```diff
@@ -46,12 +32,12 @@ public void NegativeTest(string login, string password)
             // Act
             loginPage.Login();
-            Thread.Sleep(500);
+            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => calculatorPage.IsOpened());
```
1. Please move this wait to `WaitForReady` method of CalculatorPage class.
2. Please return CalculatorPage in `Login` method. What do you think?
3. Assert calculatorPage.IsOpened(), ok?

Commit:
- Resolved ALL conversations
- Changed assertion for ValidLoginTest

varajan commented on DepositeCalcTests/Pages/CalculatorPage.cs:
```diff
@@ -112,5 +115,22 @@ public bool IsOpened()
+        {
+            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => IsOpened());
+            return new CalculatorPage(driver);
+        }
```
This page action is 'wait', it doesn't need to return new page instance.
```suggestion
        public void WeitForReady()
        {
            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => IsOpened());
        }
```

varajan commented on DepositeCalcTests/Pages/HistoryPage.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => LastCalculationsInFirstRow.Count != 0);
+            return new HistoryPage(driver);
+        }
```
```suggestion
        public void WeitForReady()
        {
            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => LastCalculationsInFirstRow.Count != 0);
        }
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,104 @@
+            calculatorPage.Calculate();
+            calculatorPage.OpenHistory();
+            var unclearedHistoryPage = historyPage.WeitForReady();
+            historyPage.Clear();
```
```suggestion
            historyPage = calculatorPage.OpenHistory();
            historyPage.WeitForReady();
            historyPage.Clear();
```

varajan commented on DepositeCalcTests/Tests/HistoryPageTests.cs:
```diff
@@ -0,0 +1,104 @@
+                calculatorPage.Calculate();
+            }
+            calculatorPage.OpenHistory();
+            historyPage.WeitForReady();
```
```suggestion
            historyPage = calculatorPage.OpenHistory();
            historyPage.WeitForReady();
```

varajan commented on DepositeCalcTests/Tests/LoginPageTests.cs:
```diff
@@ -46,15 +32,14 @@ public void NegativeTest(string login, string password)
+            CalculatorPage calculatorPage = new CalculatorPage(driver);
 
             // Act
             loginPage.Login();
```
```suggestion
            CalculatorPage calculatorPage = loginPage.Login();
```

varajan commented on DepositeCalcTests/Tests/SettingsPageTests.cs:
```diff
@@ -86,11 +69,11 @@ public void DefaultCurrencyTest(string expectedCurrency)
             var calculatorPage = new CalculatorPage(driver);
-            var settingsPage = new SettingsPage(driver);
 
             // Act
             settingsPage.Set(expectedCurrency);
             string expectedCarrencyEmblem = expectedCurrency.Substring(0, 1);
+            new WebDriverWait(driver, TimeSpan.FromSeconds(5)).Until(_ => calculatorPage.IsOpened());
```
1. Please move 'wait' to the calculator page. 
2. When you save settings, the calculator page is opened. So you can return this page at the end of the 'Set' method.
3. 'Arrange' is a good place to initialize test data/variables.
```suggestion
            // Arrange
            string expectedCarrencyEmblem = expectedCurrency.Substring(0, 1);
            
            // Act
            CalculatorPage calculatorPage = settingsPage.Set(expectedCurrency);
            calculatorPage.WaitForReady();
```

Commit:
- Resolved ALL conversations after code review
